### ISSUE TITLE
Compute controller leaks CollectionState

### ISSUE THREAD

**Issue by teskje** [MaterializeInc/materialize#15704]
### What version of Materialize are you using?

main

### How did you install Materialize?

Built from source

### What is the issue?

The compute controller never removes `CollectionState` entries it has added to `Instance::collections`, even for collections that have been dropped by the user and all active replicas.

This is a memory leak that is relatively serious, depending on how Materialize is used. `CollectionState` objects are small, but if a user creates a lot of dataflows, e.g. by issuing many queries that can't use the fast path, the memory usage will still be significant over time.

**Comment by lluki** [comment]
I have another observation that we should ensure gets fixed by this (or otherwise open a new issue with it):

Executing 
```
> CREATE TABLE t1 (f1 int);
> CREATE MATERIALIZED VIEW v1 AS SELECT MIN(f1) FROM t1; DROP MATERIALIZED VIEW v1;
```
will lead computed to forever try to fetch a snapshot that will never be available:

```
2022-11-01T14:43:16.646303Z  INFO mz_persist_client::internal::machine: snapshot sf2bd7d0d-4bff-469e-b956-6892afee97a3 as of Antichain { elements: [1667313447000] } not yet available for upper Antichain { elements: [1667313447000] } retrying in 16s
```

In this case `sf2...` is actually the shard of the materialized view, not of t1, when computed starts writing to the shard, it first fetches a snapshot of it, which fails in this case. And because we don't clean up the shard properly, this will run forever.

**Comment by vmarcos** [comment]
This bug is fixed by PR #15862. Regression testing is pending due to a dependency on relevant Prometheus metrics to monitor the size of the collection metadata in the compute controller, now tracked in #16026. Once this is resolved, the pending regression testing should be addressed in issue #16028. I am therefore closing this issue since all follow-up activities are now being tracked in separate tickets.
